### PR TITLE
Run Percy inside E2E main workflow

### PR DESCRIPTION
**Pull request by nemanjaglumac**
### Status
PENDING REVIEW

### What does this PR accomplish?
- Merges Percy workflow in E2E main workflow
- Runs **and records** Percy visual tests:
    1. on `master` only
    2. using only `EE version` of the uberjar
    3. if changed files include anything else other than Cypress files

    This was a workaround for:
    ```
    paths-ignore:
      - "frontend/test/**"
      - "!frontend/test/metabase-visual/**"
    ```
    because we cannot ignore all frontend tests in this workflow (they include Cypress).

### Possible further improvements
- We can run Percy on both `master` and the `release` branches but record only on `master`. cc @alxnddr - do you think this adds value? Remember, we are already running Percy as a sanity check `on: pull_request`. This workflow will run tests once changes are already merged.

### CODE REVIEW

File: .github/workflows/e2e-main.yml [hunk 1]
```diff
@@ -156,3 +157,61 @@ jobs:
...
+          echo "No files have been changed";
+        elif [ "$CHANGED_SPECS" == "$CYPRESS_SPECS" ]; then
+          echo "Not running Percy because the only changed files are Cypress E2E specs";
+        else
+          echo "Run Percy visual tests";
+          yarn run test-visual-run
+        fi
+      env:
+        PERCY_TOKEN: ${{ secrets.PERCY_TOKEN }}
+      if: env.PERCY_TOKEN != null && ${{ github.ref == 'refs/heads/master' }}
```
Comment (nemanjaglumac) [inline .github/workflows/e2e-main.yml:217]: @alxnddr can confirm or correct me if I'm wrong, but I understood that we should record only on `master`?

File: .github/workflows/e2e-main.yml [hunk 2]
```diff
@@ -9,6 +9,7 @@ on:
       - "docs/**"
       - "**.md"
       - ".circleci/**"
+      - "**unit.spec.js"
```
Comment (nemanjaglumac) [inline .github/workflows/e2e-main.yml:12]: Not entirely sure this glob works.
According to the [glob tester](https://toools.cloud/miscellaneous/glob-tester) it should. But IIRC I was trying it out in my fork and the workflow ran even if the only changed file was a unit test.

Comment (nemanjaglumac) [reply .github/workflows/e2e-main.yml:12]: Another potential issue with this glob is that it doesn't account for `.jsx`, `.`ts` or `.tsx` unit tests.

**Review by flamber — COMMENTED** [review]
Looks good, but will let someone else review, since I've been involved.

File: .github/workflows/e2e-main.yml (hunk 2 above)
Comment (flamber) [reply .github/workflows/e2e-main.yml:12]: Don't know much about this, but I would probably use
```suggestion
      - "**/*.unit.spec.{js,jsx,ts,tsx}"
```

File: .github/workflows/e2e-main.yml (hunk 1 above)
Comment (alxnddr) [reply .github/workflows/e2e-main.yml:217]: > @alxnddr can confirm or correct me if I'm wrong, but I understood that we should record only on `master`?

Yep, you're right. Other jobs like `percy-visual-label` will record it on branches

**Comment by diogormendes** [comment]
Just one suggestion: What do you think of adding the analysis if is needed or not to run at the begging of the workflow ? That way we can drop it as soon as we know, saving some time on `maildev`, `download artifact`, etc...

**Comment by nemanjaglumac** [comment]
> Just one suggestion: What do you think of adding the analysis if is needed or not to run at the begging of the workflow ? That way we can drop it as soon as we know, saving some time on `maildev`, `download artifact`, etc...

Just to be sure we're on the same page - the beginning of a workflow or the beginning of Percy job within this workflow?

**Comment by ariya** [comment]
> Just to be sure we're on the same page - the beginning of a workflow or the beginning of Percy job within this workflow?

Note that if the beginning of the job we want to check the diff to decide whether to continue the rest of the job or not, unfortunately GitHub Actions workflow doesn't support clean early termination. We may end up with `if`-style conditional check for subsequent steps, which IMO not much better/aesthetic than what's in this PR already.

**Comment by nemanjaglumac** [comment]
> > Just to be sure we're on the same page - the beginning of a workflow or the beginning of Percy job within this workflow?
> 
> Note that if the beginning of the job we want to check the diff to decide whether to continue the rest of the job or not, unfortunately GitHub Actions workflow doesn't support clean early termination. We may end up with `if`-style conditional check for subsequent steps, which IMO not much better/aesthetic than what's in this PR already.

Not sure if @diogormendes meant the diff check or checking if we should skip this step all together:
https://github.com/metabase/metabase/pull/23480/files#diff-42420760680849afeaba9ddbbc9c898a1bf90c5d03d7b9f8d8266513f614a5f2R217

I believe it's the later.

**Review by ariya — APPROVED** [review]

**Pull request merged**